File: macos/Services/MCP/Tools/Tier1/ListRelationshipsTool.swift
// ListRelationshipsTool.swift
// Gridex
//
// MCP Tool: List foreign key relationships for a table.

import Foundation

struct ListRelationshipsTool: MCPTool {
    let name = "list_relationships"
    let description = "List foreign key relationships for a table. Returns both incoming and outgoing references."
    let tier = MCPPermissionTier.schema

    let inputSchema: [String: Any] = [
        "type": "object",
        "properties": [
            "connection_id": [
                "type": "string",
                "description": "Connection identifier"
            ],
            "table_name": [
                "type": "string",
                "description": "Name of the table"
            ]
        ],
        "required": ["connection_id", "table_name"]
    ]

    func execute(params: JSONValue, context: MCPToolContext) async throws -> MCPToolResult {
        let connectionId = try extractConnectionId(from: params)

        guard let tableName = params["table_name"]?.stringValue else {
            throw MCPToolError.invalidParameters("table_name is required")
        }

        let permission = await context.checkPermission(tier: tier, connectionId: connectionId)
        if let error = permission.errorMessage {
            throw MCPToolError.permissionDenied(error)
        }

        let (adapter, _) = try await context.getAdapter(for: connectionId)

        // Get outgoing foreign keys (this table references other tables)
        let outgoing = try await adapter.listForeignKeys(table: tableName, schema: nil)

        // For incoming relationships, we need to scan all tables
        // This is expensive, so we'll do it if possible
        var incoming: [ForeignKeyInfo] = []
        let allTables = try await adapter.listTables(schema: nil)

        for table in allTables where table.name != tableName {
            let fks = try await adapter.listForeignKeys(table: table.name, schema: nil)
            for fk in fks where fk.referencedTable == tableName {
                incoming.append(ForeignKeyInfo(
                    name: fk.name,
                    columns: fk.columns,
                    referencedTable: table.name, // The table that references us
                    referencedColumns: fk.referencedColumns,
                    onDelete: fk.onDelete,
                    onUpdate: fk.onUpdate
                ))
            }
        }

        var result: [String: Any] = [
            "table": tableName
        ]

        // Outgoing relationships
        if !outgoing.isEmpty {
            var outList: [[String: Any]] = []
            for fk in outgoing {
                outList.append([
                    "name": fk.name ?? "",
                    "columns": fk.columns,
                    "references_table": fk.referencedTable,
                    "references_columns": fk.referencedColumns
                ])
            }
            result["outgoing"] = outList
        } else {
            result["outgoing"] = "None"
        }

        // Incoming relationships
        if !incoming.isEmpty {
            var inList: [[String: Any]] = []
            for fk in incoming {
                inList.append([
                    "from_table": fk.referencedTable,
                    "from_columns": fk.columns,
                    "to_columns": fk.referencedColumns
                ])
            }
            result["incoming"] = inList
        } else {
            result["incoming"] = "None"
        }

        let data = try JSONSerialization.data(withJSONObject: result, options: [.prettyPrinted, .sortedKeys])
        let json = String(data: data, encoding: .utf8) ?? "{}"

        return MCPToolResult(text: json)
    }
}

File: windows/Gridex/Services/MCP/Tools/Tier1/ListRelationshipsTool.cpp
//
// ListRelationshipsTool.cpp
//

#include "ListRelationshipsTool.h"
#include "../MCPToolHelpers.h"

namespace DBModels
{
    MCPToolResult ListRelationshipsTool::execute(const nlohmann::json& params, MCPToolContext& ctx)
    {
        const auto connId = MCPTool::extractConnectionId(params);

        if (!params.contains("table_name") || !params["table_name"].is_string())
            throw MCPToolError::invalidParameters("table_name is required");
        const auto tableUtf8 = params["table_name"].get<std::string>();
        const std::wstring tableW = MCPToolHelpers::fromUtf8(tableUtf8);

        const std::wstring schemaW = (params.contains("schema") && params["schema"].is_string())
            ? MCPToolHelpers::fromUtf8(params["schema"].get<std::string>()) : std::wstring{};

        const auto perm = ctx.checkPermission(tier(), connId);
        if (const auto* msg = perm.errorMessage())
            throw MCPToolError::permissionDenied(*msg);

        auto [adapter, config] = ctx.getAdapter(connId);

        nlohmann::json result;
        result["table"] = tableUtf8;

        // Outgoing: this table's FKs pointing at others.
        nlohmann::json out = nlohmann::json::array();
        for (const auto& fk : adapter->listForeignKeys(tableW, schemaW))
        {
            out.push_back({
                {"name",              MCPToolHelpers::toUtf8(fk.name)},
                {"columns",           MCPToolHelpers::toUtf8(fk.column)},
                {"references_table",  MCPToolHelpers::toUtf8(fk.referencedTable)},
                {"references_columns",MCPToolHelpers::toUtf8(fk.referencedColumn)}
            });
        }
        result["outgoing"] = out.empty() ? nlohmann::json("None") : out;

        // Incoming: walk all tables and grab FKs whose target is this
        // table. Linear in total-tables * avg-FKs-per-table; acceptable
        // for typical schemas, and the MCP call is one-shot.
        nlohmann::json in = nlohmann::json::array();
        for (const auto& t : adapter->listTables(schemaW))
        {
            if (t.name == tableW) continue;
            for (const auto& fk : adapter->listForeignKeys(t.name, schemaW))
            {
                if (fk.referencedTable == tableW)
                {
                    in.push_back({
                        {"from_table",   MCPToolHelpers::toUtf8(t.name)},
                        {"from_columns", MCPToolHelpers::toUtf8(fk.column)},
                        {"to_columns",   MCPToolHelpers::toUtf8(fk.referencedColumn)}
                    });
                }
            }
        }
        result["incoming"] = in.empty() ? nlohmann::json("None") : in;

        return MCPToolResult::text(result.dump(2));
    }
}
